### ISSUE TITLE
Please provide a way to do a dial that blocks up to some time but doesn't return error in case of timeout, or expose ConnectivityState

### ISSUE THREAD

**Issue by dzbarsky** [grpc/grpc-go#1359]
Please answer these questions before submitting your issue.

### What version of gRPC are you using?
master

### What version of Go are you using (`go version`)?
1.6

### What operating system (Linux, Windows, …) and version?
Linux

### What did you do?
Make sure server we are connecting to isn't up.
Create a client using WithBlock and WithTimeout options set.

### What did you expect to see?
Dial returns a client that does not have any servers connected and continues to watch for connection in the background. Failure occurs at RPC time if no server has been found.

Alternately we can create a client without the Block/Timeout params, but we want to be able to block in our function doing client creation until we get a connection to server (for example we could accomplish this if ConnectivityState was exposed with a callback, the way it works in Python).

### What did you see instead?
Dial fails with error, no ClientConn returned

**Comment by dzbarsky** [comment]
https://github.com/grpc/grpc-go/pull/1361 would solve my problem

**Comment by menghanl** [comment]
This can be solved by a connectivity state API. We are planning to support the API this quarter.

On the issue you mentioned here, can you give more details about why a blocking dial with timeout does not work for you?

**Comment by dzbarsky** [comment]
Let's say you do a blocking dial with a timeout and you connect to a server. By the time you do an RPC it may fail if the server has already gone down. So you must handle this case whether or not the dial succeeds.

If the initial dial times out, you get back an error and a nil ClientConn, so you have to handle this error as well. If we instead returned a ClientConn that failed at RPC time, that is fewer errors for client code to handle.

We are migrating a large set of services (100+) to gRPC, and our in-house RPC system had the behavior I am pushing for here. In particular, all call sites assume that client creation does not fail if you pass in valid arguments, and RPCs may fail if you are not connected (which is the case no matter what). At the same time, we do block for a short period of time, so if you create a client to a service that is currently available, you can immediately issue an RPC and it will work. We believe this is a better API since it makes client code simpler and makes it easier to write non-flaky tests. In order to make the migration easier (aka avoid breaking all the clients), we want to implement a "soft block" which would be a non-blocking dial that waits in our library for some timeout, but returns early if grpc ClientConn is created.

In general, ConnectivityState is what we want (as I mentioned in the issue title), but the PR is a quick bandaid that lets us get the right behavior. We are currently blocked on deploying grpc because of this issue, if ConnectivityState API is not coming soon, it would be nice to take the linked PR or something similar in the meanwhile.

**Comment by dfawley** [comment]
It sounds like what you want is to perform a non-blocking dial, but set [`grpc.FailFast(false)`](https://godoc.org/google.golang.org/grpc#FailFast) on all of your outgoing RPCs.  Would this solve the problem for you?  Note that you can use [`grpc.WithDefaultCallOptions(grpc.FailFast(false))`](https://godoc.org/google.golang.org/grpc#WithDefaultCallOptions) to avoid needing to pass the `CallOption` every time.

**Comment by dzbarsky** [comment]
No, we do not want to set FailFast(false) on our RPCs because that causes cascading slowdown failures when upstream services are down. We really do want to block on the client creation up to a specified timeout, but want to fail fast on each RPC.

**Comment by dfawley** [comment]
Would this work?

```go
ctx := context.WithTimeout(context.Background(), time.Second)
c, err := grpc.DialContext(ctx, ..., grpc.WithBlock())
if err != nil {
  c, err = grpc.Dial(...) // Not blocking
}
```

It's a bit of boilerplate, but you could factor it into a wrapper function...

We're going to be making some changes to the dialing process (and in the implementation of fail-fast AKA wait-for-ready) in the next several weeks to more closely match the specification and the C/Java implementations of gRPC, so I'm somewhat hesitant to make any other changes to it at this time, as it will only increase the chances that we will have to break something.

Note that even fail-fast (*non*-wait-for-ready) RPCs are supposed to block if the ClientConn is in the CONNECTING or IDLE state ([ref](https://github.com/grpc/grpc/blob/master/doc/wait-for-ready.md)).

**Comment by dzbarsky** [comment]
We've already tried the approach you suggested, however it doesn't work because grpc assumes ownership of the Balancer DialOpt (i.e. calls Close on it when done). We also tried doing a wrapper class around the Balancer that would block the Close, but that still doesn't work because grpc will call Start twice, which leads to data races.

We need to support our services calling into our wrapper framework with custom balancers to support special resolving logic (like cross-data center request failover, etc.)

I understand your hesitation to change this, if you have an estimate for when you can expose what we actually want here (ConnectivityState) that can help our decision whether to fork in the meantime to unblock ourselves.

**Comment by dfawley** [comment]
@MakMukhi will be working on the ConnectivityState API (#1245) and is estimating about 2 weeks for that work.

**Comment by dfawley** [comment]
Closing this in favor of #1208 (#1245 was a dupe of it).
